notebooks/tutorial_ivf_pq.ipynb

```diff
@@ -0,0 +1,1226 @@
+   "source": [
+    "## Tweaking indexing parameters\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Deciding on the indexing parameters is a bit more involved than on the search parameters. This is obviously because `ivf_pq.IndexParams` has more members than `ivf_pq.SearchParams`, but also because the try-test loop takes longer time when it includes training. Luckily training raft's ivf-pq does not take too much time thanks to using its fast implementation of Balanced Hierarchical K-Means clustering during build and an efficient encoding kernel.\n",
+    "\n",
+    "First of all, let's pick the parameters we __don't need__ to tweak:\n",
+    "\n",
+    "  - `metric` - the distance metric often depens on the problem and thus fixed (currently raft supports variations of eucliean and inner product distances)\n",
+    "  - `conservative_memory_allocation` only affects how data is allocated - does not affect the search performance\n",
+    "  - `add_data_on_build` is a convenience flag that tells if you prefer to do both training (clustering) and filling the DB in one call to `ivf_pq.build` or start with an empty DB and add all data using `ivf_pq.extend`\n",
+    "  - `force_random_rotation` may slightly affect performance when the data dimensionality is a power of two (see the module docs), but normally you don't need to change the defaults. \n",
+    "\n",
+    "The rest of the parameters can be divided in two categories: influencing the coarse search (`kmeans_n_iters`, `kmeans_trainset_fraction` , `n_lists`) and the fine search / product quantization (`codebook_kind`, `pq_dim`, `pq_bits`)."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Indexing parameters affecting the coarse search\n",
+    "\n",
+    "#### n_lists\n",
+    "\n",
+    "`n_lists` is the first parameter to look at. It has a profound impact on overall performance during both training and search.\n",
+    "`n_lists` defines the number of clusters into which the index data is partitioned; you should keep this in mind when selecting the `n_probes` search parameter.\n",
+    "\n",
+    "The ratio `n_probes/n_lists` tells how large fraction of the dataset is compared to each query. If `n_lists == n_probes`, that is like a brute force search: we compare all dataset vectors to all query vectors. One would expect the recall is equal to `1` in such a case, but that does not take into account the PQ compression; in reality the recall is always lower unless you refine the search results.\n",
+    "\n",
+    "As `n_probes` approaches `n_lists`, IVF-PQ becomes slower than brute force because of all the extra work the algorithm does: dimension padding / transform, two-step search, extra PQ compute, etc. In practice searching around 1% of lists is enough for many datasets. But this depends on how well the input can be clustered. (e.g. for uniform random numbers as inputs, IVF methods don't work well).\n",
```

Comment:
```suggestion
    "As `n_probes` approaches `n_lists`, IVF-PQ becomes slower than brute force because of all the extra work the algorithm does: dimension padding / transform, two-step search, extra PQ compute, etc. In practice searching around 0.1-1% of lists is enough for many datasets. But this depends on how well the input can be clustered. (e.g. for uniform random numbers as inputs, IVF methods don't work well).\n",
```